deeppavlov/dataset_iterators/squad_iterator.py

```diff
@@ -64,3 +66,82 @@ def _extract_cqas(data: Dict[str, Any]) -> List[Tuple[Tuple[str, str], Tuple[Lis
                         cqas.append(((context, q), (ans_text, ans_start)))
         return cqas
+
+
+@register('multi_squad_iterator')
+class MultiSquadIterator(DataLearningIterator):
+    """Dataset iterator for multiparagraph-SQuAD dataset.
+
+        With ``with_answer_rate`` rate samples context with answer and with ``1 - with_answer_rate`` samples context
+        from the same article, but without an answer. Contexts without an answer are sampled according to
+        their tfidf scores (tfidf score between question and context).
+
+        It extracts ``context``, ``question``, ``answer_text`` and ``answer_start`` position from dataset.
+        Example from a dataset is a tuple of ``(context, question)`` and ``(answer_text, answer_start)``. If there is
+        no answer in context, then ``answer_text`` is empty string and `answer_start` is equal to -1.
+
+        Args:
+            data: dict with keys ``'train'``, ``'valid'`` and ``'test'`` and values
+            seed: random seed for data shuffling
+            shuffle: whether to shuffle data during batching
+            with_answer_rate: sampling rate of contexts with answer
+
+        Attributes:
+            shuffle: whether to shuffle data during batching
+            random: instance of ``Random`` initialized with a seed
+        """
+
+    def __init__(self, data, seed: int = None, shuffle: bool = True, with_answer_rate: float = 0.666, *args, **kwargs):
+        self.with_answer_rate = with_answer_rate
+        self.seed = seed
+        self.np_random = np.random.RandomState(seed)
+        super().__init__(data, seed, shuffle, *args, **kwargs)
+
+    def gen_batches(self, batch_size: int, data_type: str = 'train', shuffle: bool = None):
+
+        if shuffle is None:
+            shuffle = self.shuffle
+
+        if data_type == 'train':
+            random = self.np_random
+        else:
+            random = np.random.RandomState(self.seed)
+
+        if shuffle:
+            random.shuffle(self.data[data_type])
+
+        data = self.data[data_type]
+        data_len = len(data)
+
+        for i in range((data_len - 1) // batch_size + 1):
+            batch = []
+            for j in range(i * batch_size, min((i+1) * batch_size, data_len)):
+                q = data[j]['question']
+                contexts = data[j]['contexts']
+                ans_contexts = [c for c in contexts if len(c['answer']) > 0]
+                noans_contexts = [c for c in contexts if len(c['answer']) == 0]
+                # sample context with answer or without answer
+                if random.rand() < self.with_answer_rate or len(noans_contexts) == 0:
+                    # select random context with answer
+                    context = random.choice(ans_contexts)
+                else:
+                    # select random context without answer
+                    # prob ~ context tfidf score
+                    noans_scores = np.array(list(map(lambda x: x['score'], noans_contexts)))
```

Comment:
please, use list comprehensions, especially, if you need a lambda
`np.array([x['score'] for x in noans_contexts])`

Reply:
fixed